contracts/chain-adapters/Linea_Adapter.sol

```diff
@@ -0,0 +1,135 @@
+     *   (BridgedToken.sol) will be automatically deployed on the target chain.
+     * @dev User should first allow the bridge to transfer tokens on his behalf.
+     *   Alternatively, you can use `bridgeTokenWithPermit` to do so in a single
+     *   transaction. If you want the transfer to be automatically executed on the
+     *   destination chain. You should send enough ETH to pay the postman fees.
+     *   Note that Linea can reserve some tokens (which use a dedicated bridge).
+     *   In this case, the token cannot be bridged. Linea can only reserve tokens
+     *   that have not been bridged yet.
+     *   Linea can pause the bridge for security reason. In this case new bridge
+     *   transaction would revert.
+     * @param _token The address of the token to be bridged.
+     * @param _amount The amount of the token to be bridged.
+     * @param _recipient The address that will receive the tokens on the other chain.
+     */
+    function bridgeToken(
+        address _token,
+        uint256 _amount,
+        address _recipient
+    ) external payable;
+}
+
+interface IUSDCBridge {
+    function usdc() external view returns (address);
+
+    /**
+     * @dev Sends the sender's USDC from L1 to the recipient on L2, locks the USDC sent
+     * in this contract and sends a message to the message bridge
+     * contract to mint the equivalent USDC on L2
+     * @param amount The amount of USDC to send
+     * @param to The recipient's address to receive the funds
+     */
+    function depositTo(uint256 amount, address to) external payable;
+}
+
+// solhint-disable-next-line contract-name-camelcase
+contract Linea_Adapter is AdapterInterface {
+    using SafeERC20 for IERC20;
+
+    WETH9Interface public immutable l1Weth;
+
+    IMessageService public immutable l1MessageService;
+    ITokenBridge public immutable l1TokenBridge;
+    IUSDCBridge public immutable l1UsdcBridge;
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1Weth WETH address on L1.
+     * @param _l1MessageService Canonical message service contract on L1.
+     * @param _l1TokenBridge Canonical token bridge contract on L1.
+     * @param _l1UsdcBridge L1 USDC Bridge to ConsenSys's L2 Linea.
+     */
+    constructor(
+        WETH9Interface _l1Weth,
+        IMessageService _l1MessageService,
+        ITokenBridge _l1TokenBridge,
+        IUSDCBridge _l1UsdcBridge
+    ) {
+        l1Weth = _l1Weth;
+        l1MessageService = _l1MessageService;
+        l1TokenBridge = _l1TokenBridge;
+        l1UsdcBridge = _l1UsdcBridge;
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Linea.
+     * @param target Contract on Linea that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        l1MessageService.sendMessage{ value: msg.value }(target, 0, message);
```

Comment:
Because Linea (currently at least) does not support auto-claiming of messages that contain calldata, we need to finalize messages ourselves. Which means we can omit fees for now. I will set the msg.value to 0

Reply:
ok and in the future if we want to omit fees we should still grab balance from address(this).balance, so that the caller doesn't have to send value on each txn